### ISSUE TITLE
Assigning to Object Series corrupts them

### ISSUE THREAD

**Issue by Wainberg** [pola-rs/polars#15294]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
>>> a = pl.Series([1], dtype=pl.Object)
>>> a[0] = 2
thread '<unnamed>' panicked at crates/polars-core/src/datatypes/dtype.rs:328:38:
called `Result::unwrap()` on an `Err` value: InvalidOperation(ErrString("cannot convert Object dtype data to Arrow"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../python3.12/site-packages/polars/series/series.py", line 1360, in __setitem__
    self.scatter(key, value)
  File ".../python3.12/site-packages/polars/series/series.py", line 4713, in scatter
    self._s.scatter(indices._s, values._s)
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: InvalidOperation(ErrString("cannot convert Object dtype data to Arrow"))
>>> a
shape: (1_830_879_488,)
Series: 'default' [null]
[
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        …
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
        null
]
>>> len(a)
0
>>> a.shape
(0,)
>>> a[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../python3.12/site-packages/polars/series/series.py", line 1327, in __getitem__
    return self._s.get_index_signed(item)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
IndexError: index 0 is out of bounds for sequence of length 0
```

### Issue description

Assigning to an object Series fails, but also corrupts the object. It appears to erase all the existing data, but prints as if it has 1.8 billion null elements.

### Expected behavior

Should behave the same as for other dtypes.

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

**Comment by itamarst** [comment]
Going to try to fix this.

**Comment by deanm0000** [comment]
@itamarst I'm curious what you're going to fix about this. A Series is immutable so it seems the only thing to fix is that it Panics although it does give InvalidOperation also.

**Comment by itamarst** [comment]
Are they?

```python
>>> a = pl.Series([1, 2, 3])
>>> a[1] = 500
>>> a
shape: (3,)
Series: '' [i64]
[
        1
        500
        3
]
```

So first, if they're supposed to be immutable then that shouldn't be possible either?

Second, even if Series are immutable: corrupting memory is _bad_. In particular, for people doing exploratory data analysis, this can destroy a whole session.

So the absolute minimum fix is "error out without corrupting memory".

**Comment by deanm0000** [comment]
I misspoke, arrow memory is immutable but polars gives utilities to recreate a series as though the underlying arrow array were mutable.


I'm guessing, then, that you're going to make those utilities (which I didn't realize existed until now) work for objects.

I wasn't trying to be argumentative, I was just really surprised that there was anything to be done except improve the error.

**Comment by itamarst** [comment]
It's a fine set of questions, I came in to this issue not knowing whether they are supposed to be mutable or not.

The reason I think it's worth addressing is that it corrupts memory, suggesting the panic is violating some `unsafe` invariant, which suggests there might be more problems lurking in there.

**Comment by itamarst** [comment]
Turns out there is no `unsafe`. Instead:

1. The `PySeries` does `std::mem::take(&mut self.series);`.
2. Because of the panic, it never does the `self.series = out` that resets it.

So the weird behavior is due to an empty default `Series`. Unclear why it has internally inconsistent state, I'll maybe look at that first.

Reading the code I found another bug: `series[[1, None]] = 5` will also wipe the series' contents, because the error path that checks for nulls (and this is not the panic path, it's different) also doesn't restore things.

**Comment by itamarst** [comment]
I guess potentially this style of bug could also occur elsewhere... anywhere there's `&mut self` method on `PySeries`. So will look and see if there are more of those and think if there's any patterns that might make it easier to avoid these sort of bugs in a systemic fashion.

**Comment by itamarst** [comment]
And another bug: out of bounds writes also corrupt the series in the same way.

**Comment by itamarst** [comment]
Going to leave this open for the fact that Object dtype Series can't set items, and do the corruption fixes in #16905.

**Comment by ritchie46** [comment]
The corruption is fixed. Assignnment of objects needs a separate feature request.
